homeassistant/components/sensor/alpha_vantage.py

```diff
@@ -80,17 +81,20 @@ def setup_platform(hass, config, add_devices, discovery_info=None):
         _LOGGER.warning(msg)
         return
 
-    timeseries = TimeSeries(key=api_key)
-
     dev = []
+
+    timeseries = TimeSeries(key=api_key)
     for symbol in symbols:
         try:
             _LOGGER.debug("Configuring timeseries for symbols: %s",
                           symbol[CONF_SYMBOL])
             timeseries.get_intraday(symbol[CONF_SYMBOL])
-        except ValueError:
+            _LOGGER.debug("Sleep %s seconds to prevent API throttling...", 10)
+            time.sleep(10)
```

Comment:
`time.sleep` is **never** allowed.

Reply:
I was already aware of that. The basic approach is 

1. Get rid of API calls at platform setup
2. Use Throttle for the update() function.

@balloob Only question is if the throttle will work for 1 specific sensor or across sensors.

Reply:
@metbril the proper way to do this is to have the main component polling & updating some shared data-structure with the polled contents. The sensors themselves are set to require no polling, and are informed by the main component when the values are updated. Unfortunately I don't know what's a simple example using this structure, only that xiaomi_aqara is doing it like that. I hope someone else will chime in on that!

Reply:
Platforms can also be based around a main data structure (client) class that interacts with an api. This structure is set up when the platform is first set up and stored in hass.data. Subsequent platform set ups should check for existing client in hass.data. The client instance pulls all data from the api in the client update method which is throttled, and stores the data in an instance attribute. Entities access the client data attribute in their update methods.